src/backend/distributed/utils/resource_lock.c

```diff
@@ -1060,3 +1086,218 @@ CitusLockTableAclCheck(Oid relationId, LOCKMODE lockmode, Oid userId)
 
 	return aclResult;
 }
+
+
+/*
+ * AcquireDistributedLockOnRelations_Internal acquire a distributed lock on worker nodes
+ * for given list of relations ids. Worker node list is sorted so that the lock
+ * is acquired in the same order regardless of which node it was run on.
+ * Notice that no lock is acquired on coordinator node if the coordinator is not
+ * added to the metadata.
+ *
+ * Notice that no validation is done on the relationIds, that is the responsibility of
+ * AcquireDistributedLockOnRelations.
+ *
+ * A nowait flag is used to require the locks to be available immediately
+ * and if that is not the case, an error will be thrown
+ */
+static void
+AcquireDistributedLockOnRelations_Internal(List *relationIdList, LOCKMODE lockMode, bool
+										   nowait)
+{
+	Oid relationId = InvalidOid;
+	List *workerNodeList = ActivePrimaryNodeList(NoLock);
+	Oid userId = GetUserId();
+
+	const char *lockModeCString = LockModeToLockModeCString(lockMode);
+
+	/*
+	 * We want to acquire locks in the same order across the nodes.
+	 */
+	workerNodeList = SortList(workerNodeList, CompareWorkerNodes);
+
+	UseCoordinatedTransaction();
+
+	int32 localGroupId = GetLocalGroupId();
+
+	foreach_oid(relationId, relationIdList)
+	{
+		char *qualifiedRelationName = generate_qualified_relation_name(relationId);
+		StringInfo lockRelationCommand = makeStringInfo();
+
+		char *lockCommand = nowait ? LOCK_RELATION_IF_EXISTS_NOWAIT :
+							LOCK_RELATION_IF_EXISTS;
+		appendStringInfo(lockRelationCommand, lockCommand,
+						 quote_literal_cstr(qualifiedRelationName),
+						 quote_literal_cstr(lockModeCString));
+
+		/* preemptive permission check so a worker connection is not */
+		/* established if the user is not allowed to acquire the lock */
+		CitusLockTableAclCheck(relationId, lockMode, userId);
+
+		WorkerNode *workerNode = NULL;
+		foreach_ptr(workerNode, workerNodeList)
+		{
+			const char *nodeName = workerNode->workerName;
+			int nodePort = workerNode->workerPort;
+
+			/* if local node is one of the targets, acquire the lock locally */
+			if (workerNode->groupId == localGroupId)
+			{
+				LockRelationIfExists(
+					cstring_to_text(qualifiedRelationName),
+					lockModeCString,
+					nowait);
+
+				continue;
+			}
+
+			SendCommandToWorker(nodeName, nodePort, lockRelationCommand->data);
```

Comment:
we typically allow users to skip sending commands to the workers when `EnableDDLPropagation` if false.

I think it would be nice to have it here as well.

Reply:
Disabling ddl prop. is a Citus internal operation for MX. Users doing so taking some risks, which we never advise (and even document).  So, we can safely ignore such cases